scipy/stats/_multivariate.py

```diff
@@ -181,6 +182,58 @@ def pinv(self):
         return self._pinv
 
 
+def _compute_sqrt_cov(cov, check_valid='warn', tol=1e-8):
+    """
+    Compute a matrix `A` such that `A @ A.T == cov` using the singular value
+    decomposition (SVD).
+
+    Parameters
+    ------
+    cov : array_like
+        Covariance matrix (2-D).
+    check_valid : {'warn', 'raise', 'ignore'}, optional
+        Behavior when `cov` is not positive semi-definite. (Default: True)
+    tol : float, optional
+        Tolerance when checking whether the matrix is positive semi-definite.
+        Note that `cov` is cast to double before checking the tolerance.
+
+    Returns
+    -------
+    numpy.ndarray
+        Matrix `A` such that `A @ A.T == cov`.
+
+    Note
+    ----
+    This function is mainly intended for use by the multivariate normal
+    distribution, or distributions which produce samples by generating a
+    multivariate normal sample as an intermediate step (e.g. multivariate t
+    distribution). Although other methods to compute the square root (such as
+    an eigenvalue or Cholesky decomposition) could potentially be faster, we
+    use the SVD to maintain backward compatibility. In effect, this method
+    performs part of the computation in
+    `numpy.random.RandomState.multivariate_normal` (which is in turn equivalent
+    to `numpy.random.Generator.multivariate_normal`).
+
+    """
+    # Cast to double to ensure `tol` remains meaningful.
+    cov = cov.astype(np.double)
+    u, s, v = np.linalg.svd(cov)
+    # Check if the matrix is positive semi-definite. If so, v.T should be close
+    # (up to round-off error) to u if the singular value of the corresponding
+    # row isn't zero. Note that this also checks whether `cov` is symmetric.
+    if check_valid not in ('warn', 'raise', 'ignore'):
+        raise ValueError("check_valid must equal 'warn', 'raise', or 'ignore'")
+    if check_valid != 'ignore':
+        is_psd = np.allclose((v.T * s) @ v, cov, rtol=tol, atol=tol)
```

Comment:
You can directly check if entries of `s` is numerically almost nonpositive. Then you don't need to upscale to floats. Symmetry shouldn't be able to arrive all the way here and should be handled somewhere before. 

```python
tol = np.max(s) * max(cov.shape) * np.finfo(cov.dtype.char.lower()).eps
assert np.all(s>tol)
```

Reply:
```
tol = np.max(s) * max(cov.shape) * np.finfo(cov.dtype.char.lower()).eps
assert np.all(s>-tol)
```

That check will always pass because the singular values are non-negative. I was actually trying to weed out non positive semi-definite matrices with this check. The symmetry check is a by-product. 

It's written this way because compatibility with numpy is paramount.

Reply:
Ah yes, my bad I was doing too much `eigh` stuff lately. It should be `+tol` not negative. numpy compatibility is important but correctness is more important.

Reply:
@ilayn Are you suggesting replacing ` is_psd = np.allclose((v.T * s) @ v, cov, rtol=tol, atol=tol)` with `np.all(s>tol)`? In other words, to _only_ check for the presence of small singular values? If so, then this simple example will fail:

```python

[ins] In [30]: cov = np.diag([1., -1.])

[ins] In [31]: (u, s, v) = np.linalg.svd(cov)

[ins] In [32]: np.all(s > np.max(s) * max(cov.shape) * np.finfo(cov.dtype.char.lower()).eps)
Out[32]: True

[ins] In [33]: np.allclose((v.T * s) @ v, cov)
Out[33]: False
```

If you're suggesting checking for numerical rank-deficiency _in addition_:

```python
is_psd = np.allclose((v.T * s) @ v, cov, rtol=tol, atol=tol) and np.all(s > tol)
```

then note that users who were trying to compute samples from near singular matrices will start seeing warnings where there were none before. If that's OK, I propose using `_eigvalsh_to_eps` here. I remember you saying there was some problem with `_eigvalsh_to_eps` (presumably the hardcoded `factor = {'f': 1E3, 'd': 1E6}`?), but that'll at least be consistent with `_PSD` until `eigvalsh_to_eps` gets fixed. 

Let me know if I understood you correctly.

Reply:
Well I am assuming the covariance array is checked before it arrives here. This is becoming a bottleneck with an SVD computation which is already expensive and an array comparison. If it is already almost positive semidefinite and we are just trying to check if it is sufficiently away from zero then we only need rank check which is what I wrote above. 

If we are instead checking everything from scratch such that the array can be anything then we need a robust and performant machinery with cholesky first and eigh as the backup. But not using the SVD hammer on a structured matrix like covariance. 

`_eigvalsh_to_eps` is now obsolete since we changed all the hardcoded values with what I wrote above.

Reply:
@ilayn The only situation where `cov` is _not_ checked before it arrives here is when the `rvs` method of a non-frozen instance is called, e.g., 

```python 
multivariate_normal.rvs(mean=np.random.rand(2), cov=np.diag([1,1]), size=100)
```

There's no way to tell what `cov` the user is going to provide, so a PSD check has to be performed from scratch. We're forced to perform an SVD because we have to maintain compatibility with `numpy`, and as far as I know, this array comparison is the only way to tell if `cov` is PSD with the SVD.

I agree that the SVD isn't optimal. I don't see how we can improve while maintaing compatibility unless we also include a `'method'` argument, like we discussed [here](https://github.com/scipy/scipy/pull/12184#discussion_r537220052). 

A more aggressive option would be to  revamp the `_PSD` class to use `cholesky` or `eigh` as you suggested, and use this PR as an opportunity to start issuing a `DeprecationWarning`, warning users that random variates generated from the same seed are about to change.

Reply:
I agree but what does numpy do so that we break compatibility? This strictly about getting the cholesky factor right because the function says sqrt but the definition says `covA @ covA.T`. If square root is needed then we should have `covA @ covA`. If just the cholesky factor then we can always fall back to SVD

```python
try:
    cholesky factorization
except LinAlgError:
    continue with eigh or svd
```

Numpy shouldn't be a problem in this flow

Reply:
> This strictly about getting the cholesky factor right? 

Well, not strictly. It's about getting a matrix `A` such that `A @ A.T == cov` for _subsequent use in multivariate normal sampling_.  The Cholesky factor happens to be one such matrix, which we can't use by default. I'll explain in a minute why. First, I want to emphasize that this function is only intended for multivariate normal sample generation, and not for general purpose factorization, as I mentioned in the `Note` associated with the function.

```
Note
----
This function is mainly intended for use by the multivariate normal
distribution, or distributions which produce samples by generating a
multivariate normal sample as an intermediate step (e.g. multivariate t
distribution). Although other methods to compute the square root (such as
an eigenvalue or Cholesky decomposition) could potentially be faster, we
use the SVD to maintain backward compatibility. In effect, this method
performs part of the computation in
`numpy.random.RandomState.multivariate_normal` (which is in turn equivalent
to `numpy.random.Generator.multivariate_normal`).
"""
```

Perhaps the function name suggests it can be used more generally. I'm happy to change it to something like  `_compute_cov_factor_for_multivariate_normal_sampling`

Why can we not use Cholesky?

Multivariate normal samples are usually generated like so:

```python
# Find A such that A @ A.T == cov
z = np.random.standard_normal()
samples = mean + A @ z
```

Also see: https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Drawing_values_from_the_distribution

A matrix `A` generated from a Cholesky factorization will lead to different `samples` than one generated from say the Eigenvalue decomposition, or the SVD, for the same `z`.

Currently in scipy, multivariate normal sample generation is delegated to numpy, which uses the SVD to compute `A`. This is why we _have_ to use the SVD to compute `A`. Otherwise, users will start seeing different samples being generated from the same seed.

>  If just the cholesky factor then we can always fall back to SVD
> 
> ```python
> try:
>     cholesky factorization
> except LinAlgError:
>     continue with eigh or svd
> ```
> Numpy shouldn't be a problem in this flow

This is where compatibility with numpy breaks. For the purposes of `multivariate_normal.rvs`, even if other methods like `eigh` or `cholesky` are applicable, the default has to remain the SVD. Perhaps `cholesky` and `eigh` can be added to this function, but I don't see them being use in the near future for multivariate normal sampling. If necessary, they can be added in a separate PR.

Reply:
Unless you're saying substitute the array check with the Cholesky factorization. I just did some tests and does seem to be faster than the array comparison.

Reply:
Yes, that's why I was persisting on using Cholesky because if that chol can be done, using anything else is just a waste of computational time because it is the standard for anything relating pos.def. matrices.

But I get your point the sample drawing part I didn't get previously. Thanks for that clarification. I think it's an oversight but at some point we might want to get rid of this since both methods span the same basis thus in the limit the results will be the same.

Let's get this as is then, after that we have to clean up some numerical stuff such as that `eigh_to_eps` etc. which is not correct anymore. These smell like legacy code anyways.